### PR TITLE
[Feat] support SP for FLUX.2-klein

### PR DESCRIPTION
**Pull request by RuixiangMa**
## Purpose
support SP (Ulysses & Ring) for FLUX.2-klein
## Test Result
* Target image:
   

* tp = 1 + 4 * nvidia 4090(24G)

`curl -s -X POST "http://localhost:8004/v1/images/edits"   -F "image=@test.jpg"   -F "prompt=Change the sky to orange sunset."   -F "guidance_scale=1.0"   -F "num_inference_steps=50"   -F "n=1"   -F "size=1024x1024"   -F "output_format=png"   | jq -r '.data[0].b64_json' | base64 --decode > output.png`

| Configuration         | Ulysses degree | Ring degree | Generation Time | Speedup | Images Generated |
|-----------------------|----------------|-------------|-----------------|---------|------------------|
| Baseline              | 1              | 1           | 25.503s         | 1.00x   |              |
| Ulysses               | 4              | 1           | 13.173s         | 1.94x   |                     |
| Ring                  | 1              | 4           | 16.866s         | 1.51x   |                   |
| Hybrid Ulysses + Ring | 2              | 2           | 14.812s         | 1.72x   |                  |

### CODE REVIEW

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py [hunk 1]
```diff
@@ -690,11 +743,27 @@ def forward(
         if txt_ids.ndim == 3:
             txt_ids = txt_ids[0]
 
-        image_rotary_emb = self.pos_embed(img_ids)
-        text_rotary_emb = self.pos_embed(txt_ids)
+        if current_omni_platform.is_npu():
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: @gcanlin do we have better ways to handle this difference? this is so awkward

Comment (hsliuustc0106) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: @wtomin PTAL

Comment (gcanlin) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: I find that there exists `npu` and `mps` hardcode in `Flux2PosEmbed`, which may be from `diffusers` library I guess. I will take a micro-refactoring PR for removing it. We could delete this npu branch temporarily if we'd like to merge this PR first.

Comment (RuixiangMa) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: > I find that there exists `npu` and `mps` hardcode in `Flux2PosEmbed`, which may be from `diffusers` library I guess. I will take a micro-refactoring PR for removing it. We could delete this npu branch temporarily if we'd like to merge this PR first.

Got it! I'll just remove this branch right now

**Comment by hsliuustc0106** [comment]
update the docs as well

**Comment by RuixiangMa** [comment]
> update the docs as well

Done

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py [hunk 2]
```diff
@@ -672,6 +706,26 @@ def forward(
 
         num_txt_tokens = encoder_hidden_states.shape[1]
 
+        sp_size = self.parallel_config.sequence_parallel_size
+        get_forward_context().sequence_parallel_size = sp_size
```
Comment (wtomin) [inline vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: To fetch the `sp_size`, using `self.parallel_config.sequence_parallel_size` would be sufficient. I don't see why you need to set `get_forward_context().sequence_parallel_size` here.

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py [hunk 3]
```diff
@@ -672,6 +706,26 @@ def forward(
...
+            hidden_states, sp_pad_size = sp_shard_with_padding(hidden_states, dim=1)
+            get_forward_context().split_text_embed_in_sp = False
+            if not hasattr(self, "_sp_forward_logged"):
+                self._sp_forward_logged = True
+                logger.info(
+                    f"[Flux2 Transformer] SP enabled: sp_size={sp_size}, world_size={sp_world_size}, "
+                    f"rank={sp_rank}, original_shape={original_shape}, chunked_shape={hidden_states.shape}"
+                )
+        else:
+            if not hasattr(self, "_sp_forward_logged"):
```
Comment (wtomin) [inline vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: Is `self._sp_forward_logged` used for debugging only?

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py (hunk 2 above)
Comment (wtomin) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: I find that `LongCatImageTransformer2DModel` also edits `get_forward_context`...

Comment (wtomin) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: `od_config = get_forward_context().omni_diffusion_config` 
`parallel_config = od_config.parallel_config`
`sequence_parallel_size = parallel_config.sequence_parallel_size`

This would be my recommendation.

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py (hunk 3 above)
Comment (RuixiangMa) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: > Is `self._sp_forward_logged` used for debugging only

yes, I'm removing it now.

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py (hunk 2 above)
Comment (RuixiangMa) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: > To fetch the `sp_size`, using `self.parallel_config.sequence_parallel_size` would be sufficient. I don't see why you need to set `get_forward_context().sequence_parallel_size` here.

ths, I keep only sp_size = self.parallel_config.sequence_parallel_size—that matches what you suggested

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py [hunk 4]
```diff
@@ -672,6 +704,12 @@ def forward(
 
         num_txt_tokens = encoder_hidden_states.shape[1]
 
+        sp_size = self.parallel_config.sequence_parallel_size
+        sp_pad_size = 0
+        if sp_size > 1:
+            hidden_states, sp_pad_size = sp_shard_with_padding(hidden_states, dim=1)
```
Comment (wtomin) [inline vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: I am a little concerned about the padding behavior here. It looks like flux2_klein does not support `hidden_states_mask` (like qwen_image), therefore, the padded tokens will participate in the attention computation, which is wrong.

If using [`Non-intrusive _sp_plan`](https://github.com/vllm-project/vllm-omni/blob/main/docs/contributing/features/sequence_parallel.md#approach-1-non-intrusive-_sp_plan-recommended), `sp_padding_size` will be automatically set in `get_forward_context()`, however, `hidden_states_mask` is required to exclude the padded tokens.

```python
# vllm_omni/diffusion/models/qwen_image/qwen_image_transformer.py
        hidden_states_mask = None  # default
        if self.parallel_config is not None and self.parallel_config.sequence_parallel_size > 1:
            ctx = get_forward_context()
            if ctx.sp_original_seq_len is not None and ctx.sp_padding_size > 0:
                # Create mask for the full (padded) sequence
                # valid positions = True, padding positions = False
                batch_size = hidden_states.shape[0]
                padded_seq_len = ctx.sp_original_seq_len + ctx.sp_padding_size
                hidden_states_mask = torch.ones(
                    batch_size,
                    padded_seq_len,
                    dtype=torch.bool,
                    device=hidden_states.device,
                )
                hidden_states_mask[:, ctx.sp_original_seq_len :] = False

```

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py (hunk 1 above)
Comment (ZJY0516) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: I remember some of these is because torch_npu doesn't support complex number

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py (hunk 4 above)
Comment (wtomin) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: @RuixiangMa I would like to invite you to join the discussion here #1324.

Comment (wtomin) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: I have two suggestions:
1. refer to #1279 , you may use `auto_pad=True` and use `hidden_states_mask` to exclude the padded tokens.
2. You may set `SequenceParallelInput(auto_pad=False)`, this will raise an error when `seq_len` not divisible by `sp_size`, which I think might be rare for Flux.2-klein. We can take care of rare cases in the future, after we make our discussion clear in #1324.

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

**Comment by wtomin** [comment]
Do you have plan to support SP with `_sp_plan` hooks?

**Comment by RuixiangMa** [comment]
> Do you have plan to support SP with `_sp_plan` hooks?

Sorry, just submitted a revision

**Comment by RuixiangMa** [comment]
> > Do you have plan to support SP with `_sp_plan` hooks?
> 
> Sorry, just submitted a revision

@wtomin Can you take a look?

**Comment by wtomin** [comment]
There is a bug related to sp in #1556, I have a bugfix in #1704. Please check if it affects your PR.

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py [hunk 5]
```diff
@@ -325,20 +384,59 @@ def forward(
...
+            forward_ctx.sp_active and not forward_ctx.split_text_embed_in_sp and text_seq_len is not None
+        )
+
+        if use_sp_single_stream and image_rotary_emb is not None:
             cos, sin = image_rotary_emb
             cos = cos.to(query.dtype)
             sin = sin.to(query.dtype)
-            query = self.rope(query, cos, sin)
-            key = self.rope(key, cos, sin)
+            txt_cos, img_cos = cos[:text_seq_len], cos[text_seq_len:]
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py:405]: are these added lines from 397 to 410 independent of any diffusion model and can be extracted?

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py [hunk 6]
```diff
@@ -705,21 +864,41 @@ def forward(
...
         concat_rotary_emb = (
-            torch.cat([text_rotary_emb[0], image_rotary_emb[0]], dim=0),
-            torch.cat([text_rotary_emb[1], image_rotary_emb[1]], dim=0),
+            torch.cat([txt_freqs_cos, img_freqs_cos], dim=0),
+            torch.cat([txt_freqs_sin, img_freqs_sin], dim=0),
         )
 
+        hidden_states_mask = None
+        ctx = get_forward_context()
+        if ctx.sp_active:
```
Comment (wtomin) [inline vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: I think `ctx.sp_active` is should not be exposed to users or developers.

In `_sp_plan` hooks design, `ctx.sp_active` is determined by `_sp_shard_depth`. `_sp_shard_depth` will be automatically handled by hook function. Therefore, no need to include  `ctx.sp_active` here.

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py (hunk 5 above)
Comment (RuixiangMa) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py:405]: I've extracted the RoPE logic into a reusable helper function(only verified flux1, flux2, and z-image so far, but the others (e.g., ovis-image) should be reusable as well.), and SP-specific logic in flux2_klein (lines 393-410) remains unchanged as it handles model-specific sequence parallelism splitting.

File: vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py (hunk 6 above)
Comment (RuixiangMa) [reply vllm_omni/diffusion/models/flux2_klein/flux2_klein_transformer.py]: > I think `ctx.sp_active` is should not be exposed to users or developers.
> 
> In `_sp_plan` hooks design, `ctx.sp_active` is determined by `_sp_shard_depth`. `_sp_shard_depth` will be automatically handled by hook function. Therefore, no need to include `ctx.sp_active` here.

 It was redundant,  I removed it

**Comment by RuixiangMa** [comment]
> There is a bug related to sp in #1556, I have a bugfix in #1704. Please check if it affects your PR.

Flux2's hook-based impl seems unaffected. I'll merge and verify

**Comment by wtomin** [comment]
I would recommend you to use `_sp_plan` hooks for sp implementation instead of intrusive modification.

Taking PR #1772  as an example, one major reason that intrusive modification is not good, it affects teacache extractor function, making it over complicated.

**Comment by RuixiangMa** [comment]
> There is a bug related to sp in #1556, I have a bugfix in #1704. Please check if it affects your PR.

merge the pr and test, it work well.

**Comment by wtomin** [comment]
Have you run offline inference with irreguar shape that involves auto padding and `hidden_states_mask`?

**Comment by RuixiangMa** [comment]
> Have you run offline inference with irreguar shape that involves auto padding and `hidden_states_mask`?

Both irreguar and offload tests passed.

**Comment by hsliuustc0106** [comment]
can you provide a detailed profiling to check whether the acceleration makes sense and what's the recommended parallelism for this model. BTW, please also check https://github.com/vllm-project/vllm-omni/issues/1806

**Comment by nuclearwu** [comment]
@RuixiangMa update doc, such as `diffusion_acceleration.md` and `parallelism_acceleration.md`

**Comment by RuixiangMa** [comment]
Done
